### DISCUSSION TITLE
List view => detail query - normalized cache not sharing data

### DISCUSSION THREAD

**Discussion by ajoslin** [urql-graphql/urql#2405] [Q&A] [answered]
I have a "list" and "detail" query to get a singular or list of items with a `my_watchlist` and `watchlist_count` variable for each:

```gql
query getStreamsWatchlist ($streamIds: [Float!]!) {
  getStreams(streamIds: $streamIds) {
    id
    my_watchlist { id }
    watchlist_count
    __typename
  }
}

query getStreamWatchlist ($streamId: Float!) {
  getStream(streamId: $streamId) {
    id
    my_watchlist { id }
    watchlist_count
    __typename
  }
}
```

Each of these return a `streams` type. 

I call the list `getStreamsWatchlist` first on load with Urql.useQuery, and it populates my graphcache fine for every list item (I checked).

However, when I then call the singular `getStreamWatchlist` for any one of these same streams (after list req/res is done), there is no data in the cache available for the stream, and it does a network request again. Whether I set the requestpolicy to 'cache-first', or 'cache-and-network'. 

I thought the idea was that the detail view wouldn't need to make this call again, but it is? And it is not recognizing the data from the list query at all.

**Comment by JoviDeCroock** [comment]
This is documented [here](https://formidable.com/open-source/urql/docs/graphcache/local-resolvers/#resolving-partial-data) TLDR we can't know that Query.getStream  resolves to an entity of Stream so we need your help for that, This is assuming you are using the normalized cache as oppposed to the default doc-cache

**Comment by ajoslin** [comment, marked as answer]
Thank you for the quick response! 

I have read that and am using the following config (which I should've attached). schema.json is generated by graphql-codegen from my backend. I also explicitly set the return types of these queries via resolvers. Am not sure if I'm doing it right for the `getStreams` one though?

```js
import { dedupExchange, fetchExchange, createClient } from 'urql'
import { cacheExchange } from '@urql/exchange-graphcache'

const token = getPersistedUserToken()

export const urqlCache = cacheExchange({
  schema,
  updates: {
    Query: {}
  },
  resolvers: {
    Query: {
      getStreams: (result, args, cache, info) => {
        return args.streamIds.map((item) => ({
          __typename: 'streams',
          id: item
        }))
      },
      getStream: (result, args, cache, info) => {
        // if (parent.getStream) debugger
        return { __typename: 'streams', id: args.id }
      }
    }
  }
})

export const urqlClient = createClient({
  url: process.env.REACT_APP_API_URL + '/next/graphql',
  exchanges: [dedupExchange, urqlCache, fetchExchange],
  requestPolicy: 'cache-and-network',
  fetchOptions: () => ({
    headers: token ? { authorization: `Bearer ${token}` } : {}
  })
})
```

**Reply by kitten** [reply]
I'm not sure if this is your example and just a typo or your actual issue, but in your resolver you're using `args.id`, however, in your query you're showing that the argument is called `streamId`

**Reply by ajoslin** [reply]
Thank you @kitten! 

I fixed the `args.id` to say `args.streamId` instead, and I removed my `resolvers.Query.getStreams` then it worked.

I guess I don't need resolvers.Query.getStreams anyway because it figures out the type as soon as the response arrives.

**Comment by ajoslin** [comment]
Shouldn't schema introspection remove the need for setting `resolver.Query` hints? I'm using a schema generated by https://www.graphql-code-generator.com/plugins/urql-introspection.

I realize the docs don't say it does, but it seems like the schema introspection data would have enough info to let the graph know that getStream returns a `streams`.

**Comment by ajoslin** [comment]
Do you all know if there's a way to configure graphql-codegen to generate a resolvers file for straightforward resolvers like these?
